File: src/strands/experimental/checkpoint/checkpoint.py
"""Checkpoint system for durable agent execution.

Checkpoints capture agent state at cycle boundaries so a durability provider
(e.g. Temporal) can persist them and resume after failures.

Positions per ReAct cycle:
- ``after_model``: model call completed, tools not yet executed.
- ``after_tools``: all tools executed, next model call pending.

Per-tool granularity within a cycle is the ToolExecutor's responsibility
(e.g. TemporalToolExecutor routes each tool to a separate activity).

Usage (mirrors the interrupt pattern):
- Pause: ``AgentResult`` with ``stop_reason="checkpoint"`` and a populated
  ``checkpoint`` field.
- Resume: pass ``[{"checkpointResume": {"checkpoint": checkpoint.to_dict()}}]``
  as the next prompt.

Precedence:
- Interrupts > checkpoint: an interrupt raised during a checkpointing cycle
  returns ``stop_reason="interrupt"`` and skips the ``after_tools`` checkpoint.
- Cancel > checkpoint: a cancel signal set at either checkpoint boundary
  suppresses emission and surfaces as ``stop_reason="cancelled"``.

Known limitations:
- ``EventLoopMetrics`` resets per invocation; aggregate across resumes yourself.
- ``OpenAIResponsesModel(stateful=True)`` is not supported — the server-side
  ``response_id`` is not captured.
- At ``after_tools``, ``AgentResult.message`` is the assistant's tool-use
  message; tool results live in the snapshot.
- ``BeforeInvocationEvent`` / ``AfterInvocationEvent`` fire on every resume,
  same as interrupts.
"""

import logging
from dataclasses import asdict, dataclass, field
from typing import Any, Literal

from ...types.exceptions import CheckpointException

logger = logging.getLogger(__name__)

CHECKPOINT_SCHEMA_VERSION = "1.0"

CheckpointPosition = Literal["after_model", "after_tools"]


@dataclass(frozen=True)
class Checkpoint:
    """Pause point in the agent loop. Treat as opaque — pass back to resume.

    Attributes:
        position: What just completed (``after_model`` or ``after_tools``).
        cycle_index: ReAct loop cycle (0-based).
        snapshot: Serialized agent state from ``Snapshot.to_dict()``. Stored as
            a dict (not a ``Snapshot``) so the checkpoint is JSON-serializable.
        app_data: Opaque application-level state. The SDK does not read or
            modify this. Distinct from ``Snapshot.app_data`` (agent-level).
        schema_version: Used to reject incompatible checkpoints on resume.
    """

    position: CheckpointPosition
    cycle_index: int = 0
    snapshot: dict[str, Any] = field(default_factory=dict)
    app_data: dict[str, Any] = field(default_factory=dict)
    schema_version: str = field(init=False, default=CHECKPOINT_SCHEMA_VERSION)

    def to_dict(self) -> dict[str, Any]:
        """Serialize for persistence."""
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "Checkpoint":
        """Reconstruct from a dict produced by to_dict().

        Args:
            data: Serialized checkpoint data.

        Raises:
            CheckpointException: If schema_version doesn't match the current version.
        """
        version = data.get("schema_version", "")
        if version != CHECKPOINT_SCHEMA_VERSION:
            raise CheckpointException(
                f"Checkpoints with schema version {version!r} are not compatible "
                f"with current version {CHECKPOINT_SCHEMA_VERSION}."
            )
        known_keys = {k for k in cls.__dataclass_fields__ if k != "schema_version"}
        unknown_keys = set(data.keys()) - known_keys - {"schema_version"}
        if unknown_keys:
            logger.warning("unknown_keys=<%s> | ignoring unknown fields in checkpoint data", unknown_keys)
        return cls(**{k: v for k, v in data.items() if k in known_keys})

File: strands-ts/src/experimental/checkpoint.ts
/**
 * Checkpoint system for durable agent execution.
 *
 * **Experimental** — this module is experimental and subject to change in future
 * revisions without notice.
 *
 * A {@link Checkpoint} is a pause-point marker emitted at agent cycle boundaries.
 * It captures the position (which boundary fired) and the cycle index. It does
 * **not** capture conversation state — pair with a `SessionManager` for
 * cross-process state continuity.
 *
 * Positions per ReAct cycle:
 * - `after_model`: model returned tool use; tools have not run yet.
 * - `after_tools`: tools finished; the next model call has not happened yet.
 *
 * Per-tool granularity within a cycle is the tool executor's responsibility.
 *
 * Usage (mirrors interrupts):
 * - Pause: `AgentResult` with `stopReason === 'checkpoint'` and `checkpoint` populated.
 * - Resume: pass back `{ checkpointResume: { checkpoint: ckpt.toJSON() } }`.
 *
 * Precedence:
 * - Interrupt takes precedence over checkpoint: an interrupt during a
 *   checkpointing cycle returns `stopReason === 'interrupt'` and skips `after_tools`.
 * - Cancel takes precedence over checkpoint: a cancel signal at either boundary
 *   returns `stopReason === 'cancelled'`.
 *
 * Notes:
 * - Checkpoints are only emitted on tool-use cycles. A turn with no tool calls
 *   emits no checkpoint; use a `SessionManager` for durability of every turn.
 * - Metrics reset per invocation; aggregate yourself if needed.
 * - `BeforeInvocationEvent` / `AfterInvocationEvent` fire on every resume,
 *   same as interrupts.
 */

import { logger } from '../logging/logger.js'
import { CheckpointError } from '../errors.js'

/**
 * Current checkpoint schema version. Bumped when the serialized shape changes
 * in a way that is not backward compatible; {@link Checkpoint.fromJSON} rejects
 * checkpoints carrying any other version.
 *
 * @experimental
 */
export const CHECKPOINT_SCHEMA_VERSION = '1.0'

/**
 * Which cycle boundary a checkpoint was emitted at.
 * - `after_model`: model returned tool use; tools have not run yet.
 * - `after_tools`: tools finished; the next model call has not happened yet.
 *
 * @experimental
 */
export type CheckpointPosition = 'after_model' | 'after_tools'

/**
 * Serialized form of a {@link Checkpoint}, as produced by {@link Checkpoint.toJSON}
 * and accepted by {@link Checkpoint.fromJSON}.
 *
 * @experimental
 */
export interface CheckpointData {
  position: CheckpointPosition
  cycleIndex?: number
  /** Populated by {@link Checkpoint.toJSON}. {@link Checkpoint.fromJSON} requires it to match the current version. */
  schemaVersion?: string
}

/**
 * Resume payload passed back to a checkpointing agent to continue a durable run.
 * The whole invocation argument is this object; the agent consumes it, restores
 * its cycle position, and continues without appending any new input messages.
 * Resume with `{ checkpointResume: { checkpoint: ckpt.toJSON() } }`.
 *
 * @experimental
 */
export interface CheckpointResumeContent {
  checkpointResume: { checkpoint: CheckpointData }
}

/**
 * Pause-point marker. Treat as opaque — pass it back to resume.
 *
 * @experimental
 */
export class Checkpoint {
  /** Which boundary fired (`after_model` or `after_tools`). */
  readonly position: CheckpointPosition

  /** ReAct loop cycle (0-based). */
  readonly cycleIndex: number

  /**
   * Schema version. Always set to {@link CHECKPOINT_SCHEMA_VERSION}; not accepted
   * from the constructor so it cannot be forged. Rejects incompatible checkpoints
   * on {@link Checkpoint.fromJSON}.
   */
  readonly schemaVersion: string = CHECKPOINT_SCHEMA_VERSION

  constructor(data: { position: CheckpointPosition; cycleIndex?: number }) {
    this.position = data.position
    this.cycleIndex = data.cycleIndex ?? 0
  }

  /**
   * Serializes the checkpoint for persistence.
   */
  toJSON(): Required<CheckpointData> {
    return {
      position: this.position,
      cycleIndex: this.cycleIndex,
      schemaVersion: this.schemaVersion,
    }
  }

  /**
   * Reconstructs a checkpoint from a value produced by {@link Checkpoint.toJSON}.
   * Round-trips cleanly: `Checkpoint.fromJSON(ckpt.toJSON())` type-checks and
   * returns an equivalent checkpoint.
   *
   * @param data - Serialized checkpoint data.
   * @throws CheckpointError If `schemaVersion` does not match the current version.
   */
  static fromJSON(data: CheckpointData): Checkpoint {
    const version = data.schemaVersion ?? ''
    if (version !== CHECKPOINT_SCHEMA_VERSION) {
      throw new CheckpointError(
        `Checkpoints with schema version ${JSON.stringify(version)} are not compatible ` +
          `with current version ${CHECKPOINT_SCHEMA_VERSION}.`
      )
    }

    const knownKeys = new Set(['position', 'cycleIndex', 'schemaVersion'])
    const unknownKeys = Object.keys(data).filter((key) => !knownKeys.has(key))
    if (unknownKeys.length > 0) {
      logger.warn(`unknown_keys=<${unknownKeys.join(', ')}> | ignoring unknown fields in checkpoint data`)
    }

    return new Checkpoint({
      position: data.position,
      ...(data.cycleIndex !== undefined && { cycleIndex: data.cycleIndex }),
    })
  }
}
